src/macros-by-example.adoc

```diff
@@ -0,0 +1,823 @@
+language treats symbols as variable references.  Here, `a` and `c` in the template refer to the
+parameters of the macro, and during expansion they are “filled in” with the values supplied by
+the invocation of the macro.
+
+The template language also treats lists quasi-literally, and every element inside the list is an
+expression.  Here’s a silly macro to illustrate:
+
+[{mrk}]
+----
+(*public* reverse [a, b] [b, a])
+----
+----
+(:reverse first {amount:a, currency:c}) ⇒ [{amount:a, currency:c}, first]
+----
+
+Note that the `a` in the E-expression is _not_ part of the
+expression language and not a reference to the macro’s first parameter or any other named entity.
+From the point of view of ``reverse``’s template, the input struct is literal data.
+
+E-expressions can nest, so we could also encode the same data using `price`:
+
+----
+(:reverse first (:price a c)) ⇒ [{amount:a, currency:c}, first]
+----
+
+
+=== Invoking Macros from Templates
+
+So the template language treats symbols as parameter references, other scalars as literals, and
+structs and lists as quasi-literals, but what about S-expressions?
+
+[{mrk}]
+----
+(*public* website_url
+  [path]
+  (make_string "https://www.amazon.com/" path))
+----
+
+Template expressions that are S-expressions are operator invocations, where the operators are
+either special forms or macro invocations.  In this case, it’s the latter: an invocation of the
+system macro (_i.e._, built-in function) `make_string`, which concatenates its arguments to produce
+a single string:
+
+----
+(:website_url "gp/cart") ⇒ "https://www.amazon.com/gp/cart"
+----
+
+In the template language, macro invocations can appear almost anywhere:
+
+[{mrk}]
+----
+(*public* detail_page_url
+  [asin]
+  (website_url (make_string "dp/" asin)))
+----
+----
+(:detail_page_url "B08KTZ8249") ⇒ "https://www.amazon.com/dp/B08KTZ8249"
+----
+
+NOTE: While this doesn’t look like much of an improvement, the full string takes 38 bytes to encode,
+but the macro invocation takes as few as 12 bytes.
+
+
+=== Special Form: `*literal*`
+
+When a template-expression is syntactically an S-expression, its first
+element must be a symbol that matches either a set of keywords denoting the special forms, or the
+name of a previously-defined macro.  The interpretation of the S-expression’s remaining elements
+depends on how the symbol resolves.  In the case of macro invocations, we’ve seen above that the
+following elements are (so far!) arbitrary template expressions, but for special forms that’s not
+always the case. The `*literal*` form makes this clear:
+
+[{mrk}]
+----
+(*public* USD_price [dollars] (price dollars (*literal* USD)))
+----
+----
+(:USD_price 12.99) ⇒ { amount: 12.99, currency: USD }
+----
+
+In this template, we can’t just write `(price dollars USD)` because the symbol `USD` would be
+treated as an unbound variable reference and a syntax error, so we turn it into literal data by
+“escaping” it with `*literal*`.
```

Comment:
Can we turn expansion back on with `(unquote ..` I mean `(unliteral ..`? 😜

Reply:
No, this isn't `quasiquote`.  I wasn't planning to add that since I think the existing forms are sufficient for most use cases. The quasiliteral list/strict behavior helps quite a bit.  Also `quasiquote/unquote` is subtle and tricky to get right.